### PR TITLE
bacpop-148 UI rewrite

### PR DESCRIPTION
**Pull request by absternator**
This is a feature branch for the new UI rewrite. All consequent changes will be merged into this branch and when ui-rewrite is done this can be merged.  The styles and theming can/will be changed in subsequent PRs as the UI becomes more complete

The stack for the new UI is : ts, Vue, vite, pinia, primevue for styling, playwright & jest 
Current functionality done 
- login and loguts 
- setup of app + theming
- home page 
- routing 

To run and dev test have backend running and server on localhost 4000... change the `client_url` to `http://localhost:5173` in `app/server/src/resources/config.json`. Then go into app/client-v2 and run `npm run dev`. The app is now avaible on http://localhost:5173

image
image
image

### CODE REVIEW

File: app/client-v2/src/views/HomeView.vue [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+    })
+      .post({ name: newData.name })
+      .json();
+    if (error.value) {
+      showErrorToast('Error renaming project');
+      return;
+    }
+    data.value.data[index] = newData;
+    showSuccessToast('Project renamed successfully');
+    refetchProjects();
```
Comment (EmmaLRussell) [inline app/client-v2/src/views/HomeView.vue:65]: Do you need to both update the appropriate row, and also refetch projects? Could just have one or the other couldn't you?

**Review by EmmaLRussell — COMMENTED** [review]
I can't run this - there doesn't seem to be a package.json file committed for client-v2. 

> To run and dev test have backend running and server on localhost 4000... change the client_url to http://localhost:5173 in app/server/src/resources/config.json. Then go into app/client-v2 and run npm run dev. The app is now avaible on http://localhost:5173/

Nice idea to leave the old client in place while developing the new one, but if the plan is not to merge this to main until the whole UI is done, it feels like you might as well update the run scripts and config for the new front end in this feature branch - it will get annoying to have to keep changing those! It will mean that you'll be starting from ground zero with the tests and CI workflow, but I think that's appropriate. It doesn't look like the git workflow has changed, so presumably CI is still running the tests (unit and e2e) and linting for the old front end. I think we should get those switched over asap on this branch (though could be in another PR).

> jest

Looks more like vitest to me! :)

I know you're planning on changing the theme, but could you just make a note somewherein the README of where the current theme came from (PrimeVue I think). and how to change it - do you need to run something on the command line to get PrimeVue to pull a different theme's assets into the public folder? (Oh, or is that just a manual copy? https://primevue.org/theming/)

On log2o.svg - I think in general, it's better not to embed text in logos for accessibility particularly if they're being used as links. Should be ok if we include appropriate alt text. But might want to consider just using in page text.

**Comment by EmmaLRussell** [comment]
Thanks! I can run it now, looks great. 

Yeah, I think when you create a new project, it should open it immediately. 

I'd capitalise 'logout' and probably wouldn't bother making it red - it's a bit harder to read on the dark theme,

File: .vscode/launch.json [hunk 2]
```diff
@@ -0,0 +1,15 @@
+{
```
[the remark this answers is not in the document]
Comment (absternator) [reply .vscode/launch.json:1]: woops have added to gitignore

File: app/client-v2/.eslintrc.cjs [hunk 3]
```diff
@@ -0,0 +1,15 @@
+/* eslint-env node */
+require('@rushstack/eslint-patch/modern-module-resolution')
+
+module.exports = {
+  root: true,
+  'extends': [
+    'plugin:vue/vue3-essential',
+    'eslint:recommended',
+    '@vue/eslint-config-typescript',
+    '@vue/eslint-config-prettier/skip-formatting'
```
[the remark this answers is not in the document]
Comment (absternator) [reply app/client-v2/.eslintrc.cjs:10]: it was the default but yeah seems like it... sweet as will add that in

File: app/client-v2/.prettierrc.json [hunk 4]
```diff
@@ -0,0 +1,8 @@
+{
+  "$schema": "https://json.schemastore.org/prettierrc",
+  "semi": true,
+  "tabWidth": 2,
+  "singleQuote": true,
+  "printWidth": 100,
```
[the remark this answers is not in the document]
Comment (absternator) [reply app/client-v2/.prettierrc.json:6]: updated as requested

File: app/client-v2/README.md [hunk 5]
```diff
@@ -0,0 +1,71 @@
+# client-v2
```
[the remark this answers is not in the document]
Comment (absternator) [reply app/client-v2/README.md:1]: slimmed down

File: app/client-v2/components.d.ts [hunk 6]
```diff
@@ -0,0 +1,32 @@
+/* eslint-disable */
+/* prettier-ignore */
+// @ts-nocheck
+// Generated by unplugin-vue-components
+// Read more: https://github.com/vuejs/core/pull/3399
+export {}
+
+declare module 'vue' {
+  export interface GlobalComponents {
```
[the remark this answers is not in the document]
Comment (absternator) [reply app/client-v2/components.d.ts:9]: generated file from `unplugin-vue-components` thaat allows auto import of primevue components

File: app/client-v2/tsconfig.app.json [hunk 7]
```diff
@@ -0,0 +1,15 @@
+{
```
[the remark this answers is not in the document]
Comment (absternator) [reply app/client-v2/tsconfig.app.json:1]: its just the default with install

File: app/client/public/worker.js [hunk 8]
```diff
@@ -1,37 +1,36 @@
-if ('function' === typeof importScripts) {
-  importScripts('./web_amr_prediction.js');
-  importScripts('./web_sketch.js');
+if ("function" === typeof importScripts) {
```
[the remark this answers is not in the document]
Comment (absternator) [reply app/client/public/worker.js:1]: unsure why this happened but have now updated

File: app/client-v2/src/layouts/AppNav.vue [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+import { RouterLink } from 'vue-router';
+import { computed, ref } from 'vue';
+import { getApiUrl } from '@/config';
+import { useUserStore } from '@/stores/userStore';
+import { usePrimeVue } from 'primevue/config';
+import { useStorage } from '@vueuse/core';
+import { useTheme } from '@/composables/useTheme';
+
+const { toggleTheme, themeIcon } = useTheme();
+const PrimeVue = usePrimeVue();
```
[the remark this answers is not in the document]
Comment (absternator) [reply app/client-v2/src/layouts/AppNav.vue:11]: good catch removed

File: app/client-v2/src/stores/userStore.ts [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+  getters: {
+    isAuthenticated(state) {
+      return !!state.id;
+    }
+  },
+  actions: {
+    // todo: proper error handling
+    async getUser() {
+      try {
+        const res = await fetch(`${getApiUrl()}/user`, {
```
[the remark this answers is not in the document]
Comment (absternator) [reply app/client-v2/src/stores/userStore.ts:29]: i dont see the need to replace fetch but if you had any concerns we can discuss and make ticket accordingly

File: app/client-v2/src/views/AboutView.vue [hunk 11]
```diff
@@ -0,0 +1,20 @@
+<script setup lang="ts">
+import { ref } from 'vue'
+const date = ref(new Date())
+</script>
+
+<template>
+  <div class="p-5 mb-3 border-round-sm surface-card">
+    <Calendar v-model="date" />
```
[the remark this answers is not in the document]
Comment (absternator) [reply app/client-v2/src/views/AboutView.vue:8]: haha was just testing.. removed now

File: app/client-v2/src/views/HomeView.vue [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+const showErrorToast = (msg: string) => {
+  toast.add({ severity: 'error', summary: 'Error', detail: msg, life: 3000 });
+};
+const apiUrl = getApiUrl();
+const {
+  data,
+  error: projectsError,
+  isFetching,
+  execute: refetchProjects
+} = useFetch(apiUrl + '/projects', { credentials: 'include' }).json<ProjectsResponse>();
```
[the remark this answers is not in the document]
Comment (absternator) [reply app/client-v2/src/views/HomeView.vue:25]: yup as this is the only page that needs to know this data and not a global thing... in terms of mocking `msw` like packit for mocking apis... and generic error handling i could create generic error component or something

File: app/client-v2/src/views/HomeView.vue [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+  })
+    .post({ name: newProjectName.value })
+    .json();
+
+  if (error.value) {
+    showErrorToast('Error creating project');
+    return;
+  }
+  newProjectName.value = '';
+  showSuccessToast('Project created successfully');
```
[the remark this answers is not in the document]
Comment (absternator) [reply app/client-v2/src/views/HomeView.vue:45]: yup will remove and make creating new to navigate to new project page

**Review by absternator — COMMENTED** [review]
> Thanks! I can run it now, looks great. 

Yeah, I think when you create a new project, it should open it immediately. 

I'd capitalise 'logout' and probably wouldn't bother making it red - it's a bit harder to read on the dark theme,  

Yup will open immediatly and logout is capital but have made shade lighter red so easier to see in dark mode. As i think should still be red to give warning to users

File: app/client-v2/src/views/HomeView.vue (hunk 1 above)
Comment (absternator) [reply app/client-v2/src/views/HomeView.vue:65]: thats true have updaetd to just refetch

File: app/client-v2/src/views/HomeView.vue [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+              label="Clear"
+              outlined
+              class="ml-2"
+              v-if="filters.name.value"
+              @click="filters.name.value = null"
+            />
+          </span>
+          <div>
+            <InputGroup>
+              <InputText placeholder="Create new Project" v-model="newProjectName" @keydown.enter="addProject" />
```
Comment (M-Kusumgar) [inline app/client-v2/src/views/HomeView.vue:114]: just some consistent casing
```suggestion
              <InputText placeholder="Create new project" v-model="newProjectName" @keydown.enter="addProject" />
```

**Review by M-Kusumgar — COMMENTED** [review]
looks really good man! i like the theme, couple minor comments, but would be nice if the login page didnt have a scroll and just fit the screen, probably some extra padding or margin somewhere.

Also not sure about the calendar either on the about page haha

The text in the Beebop logo is not vertically aligned to the rest of the navbar test, wonder if we should include that text as html instead of baked into the logo?

I can make multiple projects called the same thing, i am assuming this was the case in beebop originally?

**Pull request closed without merging**